app/assets/javascripts/discourse/app/components/bookmark.js

```diff
@@ -0,0 +1,394 @@
+import {
+  LATER_TODAY_CUTOFF_HOUR,
+  MOMENT_THURSDAY,
+  laterToday,
+  now,
+  parseCustomDatetime,
+  startOfDay,
+  tomorrow,
+} from "discourse/lib/timeUtils";
+
+import { AUTO_DELETE_PREFERENCES } from "discourse/models/bookmark";
+import Component from "@ember/component";
+import I18n from "I18n";
+import KeyboardShortcuts from "discourse/lib/keyboard-shortcuts";
+import { Promise } from "rsvp";
+import { TIME_SHORTCUT_TYPES } from "discourse/lib/timeShortcut";
+
+import { action } from "@ember/object";
+import { ajax } from "discourse/lib/ajax";
+import bootbox from "bootbox";
+import discourseComputed, { on } from "discourse-common/utils/decorators";
+import { formattedReminderTime } from "discourse/lib/bookmark";
+import { and, notEmpty, or } from "@ember/object/computed";
+import { popupAjaxError } from "discourse/lib/ajax-error";
+
+// global shortcuts that interfere with these modal shortcuts, they are rebound when the
+// modal is closed
+//
+// d deletePost
+const GLOBAL_SHORTCUTS_TO_PAUSE = ["d"];
+const BOOKMARK_BINDINGS = {
+  enter: { handler: "saveAndClose" },
+  "d d": { handler: "delete" },
+};
+
+export default Component.extend({
+  tagName: "",
+
```

Comment:
I think we lack a lot of properties which should be set to null here no?

Reply:
Why do they need to be set to null here when they are being set to null in `init` anyway?

Reply:
This is the way

Reply:
More seriously I don't have a very precise answer, @eviltrout has always asked me to do for perf reasons and the documentation is always doing it this way. I imagine it's related to the way these are long lived singletons.
Or maybe it's just faster to create an object with these properties defined this way instead of waiting for init to kick?

Reply:
Right that makes sense, I will set them to null, I kind of did it in this PR a different way to see if it was wrong or OK lol.

Reply:
The reason is you get better performance if you don't change the shape of an object.

The short reason why is that javascript engines recognize common "shapes" of objects (exact sets of properties) and re-use them via caching for high performance. When the shape of an object changes (ie adding/deleting a property) they can't use those same caches and have to create a new cache or other mechanism. By establishing your properties up front, that means init does not change the shape of the object and helps the engine a bit.

(see https://mathiasbynens.be/notes/shapes-ics for a longer explanation.)

To be honest the performance difference is probably minimal but it's a good standard to stick to. Also it means when you open a class you can what properties it should have up top.

Reply:
@eviltrout thanks a lot for that answer, I will have a read of the linked article!